tests/test_tree_kem_kat.rs

```diff
@@ -0,0 +1,278 @@
+        .unwrap();
+
+    setup
+        .distribute_to_members(&adder.identity, group, &messages)
+        .unwrap();
+
+    setup.deliver_welcome(welcome.unwrap(), group).unwrap();
+
+    let addee_groups = addee.groups.borrow();
+    let addee_group = addee_groups.get(&group_id).unwrap();
+
+    let path_secrets = addee_group.export_path_secrets();
+
+    let root_secret_after_add = path_secrets.last().unwrap().clone();
+    let my_path_secret = path_secrets.first().unwrap().clone();
+
+    drop(path_secrets);
+
+    let ratchet_tree_extension_before =
+        RatchetTreeExtension::new(addee_group.export_ratchet_tree()).to_extension_struct();
+    let ratchet_tree_before = ratchet_tree_extension_before.extension_data();
+
+    let tree_hash_before = addee_group.tree_hash();
+
+    drop(addee_group);
+    drop(addee_groups);
+    drop(addee);
+
+    let mut updater_id = group.random_group_member();
+    while updater_id == addees[0] {
+        updater_id = group.random_group_member();
+    }
+
+    let (updater_index, _) = group
+        .members
+        .iter()
+        .find(|(_, id)| id == &updater_id)
+        .unwrap()
+        .clone();
+
+    let updater = clients.get(&updater_id).unwrap().borrow();
+    let mut updater_groups = updater.groups.borrow_mut();
+    let updater_group = updater_groups.get_mut(&group_id).unwrap();
+    let group_context = updater_group.export_group_context().serialized().to_vec();
+
+    let (messages, _) = updater_group.self_update(&updater.key_store, None).unwrap();
+
+    let update_path = match messages.first().unwrap() {
+        MlsMessage::Plaintext(pt) => match pt.content() {
+            MlsPlaintextContentType::Commit(commit) => commit.path().as_ref().unwrap().clone(),
+            _ => panic!("The message should not be anything but a commit."),
+        },
+        _ => panic!("The message should not be a ciphertext."),
+    };
+
+    // Drop all the borrows as not to cause problems when having the setup
+    // distribute to members.
+    drop(updater_group);
+    drop(updater_groups);
+    drop(updater);
+    drop(adder);
+    drop(clients);
+
+    setup
+        .distribute_to_members(&updater_id, group, &messages)
+        .unwrap();
+
+    // The update was sent, now we get the right state variables again
+    let clients = setup.clients.borrow();
+    let addee = clients.get(&addees[0]).unwrap().borrow();
+    let addee_groups = addee.groups.borrow();
+    let addee_group = addee_groups.get(&group_id).unwrap();
+    let tree = addee_group.export_ratchet_tree();
+
+    let my_key_package_after = tree
+        .iter()
+        .find(|node_option| {
+            if let Some(node) = node_option {
+                if let Some(key_package) = node.key_package() {
+                    if key_package.credential().identity() == &addees[0] {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        })
+        .unwrap()
```

Comment:
There are a lot of unwraps here... not sure if they are all necessary.

Reply:
It's just lots of option types to dig through until I reach my own key package. It's going to be fewer after the tree rewrite ;-)